### PR TITLE
WIP: Support real integration testing

### PR DESCRIPTION
**Pull request by per-samuelsson**
As issued in #9.

I'm still not sure how we should push this thing, since all options seem, well, not great. 

To do real integration testing of these extensions, we need to either duplicate all testing utilities we have in Nova, or expose them somehow, to tests we have here. Right now, I locally just reference the Nova sources, and have switched focus to write tests on nested transactions, including the incorporated Rollback() functionality.

For now, I've pushed a temporary commit, so if you put this repo as a sibling of `Starcounter.Nova`, it should work to run the integration tests. We shouldn't merge that though.

Please review and test NestedTransactorTests mainly. Once we have proper testing of that in isolation, I'll extend multi-decorator tests with similar capabilities.

(Also note that this is branched off of `issue/6`, which is the implementation of nested transactions in the first place. I'll probably merge them in order once we see integration tests work well).

### CODE REVIEW

**Comment by miyconst** [comment]
I am going to go through each test and see what and how it does, regarding the testing capabilities, I think we can do the following:

- Move all shared testing features into a project, let's say: `Starcounter.Database.Nova.TestResources`.
- This project will be packed into it's own NuGet package.
- This way we can consume the same testing capabilities internally and externally for extension libraries or customer projects.

**Comment by per-samuelsson** [comment]
> * Move all shared testing features into a project, let's say: `Starcounter.Database.Nova.TestResources`.
> * This project will be packed into it's own NuGet package.

Haha, we [had _exactly_ that](https://github.com/Starcounter/Starcounter.Nova/tree/84b198c55a02b365a51e8128368f59cc45dd85f5/test/Starcounter.Database.TestResources), even named `.TestResources` some time ago, for the reason that we wanted to give ability to app developers to easily write integration tests like this. 😄 

I'm redoing it now, but will trim the initial API down to a minimum.

**Comment by miyconst** [comment]
>Haha, we had *exactly* that, even named `.TestResources` some time ago ...

And that's where I took the name for the proposition 😄.

**Comment by per-samuelsson** [comment]
@miyconst 

Like recent months, I've noticed we have more and more commits where message is just `Update [File].cs`. I think we should try to stay away from that. Even if we just fix whitespacing, it's better to know it's that - "Fixed whitespace" - because then you can ignore such commits, when you are looking for a certain one.

Even when there is a suggested commit, you can still write a custom message when you go ahead and accept it.

Other than that - nice improvements. 👍

**Comment by miyconst** [comment]
@per-samuelsson, agreed. My only "excuse" for the default commit messages was the fact that we will squash and merge everything as a single commit, but I do agree that it's not the best idea.

**Comment by per-samuelsson** [comment]
With latest commits, I get tests to pass based on packages built from `develop-inner-transact`, so if you want to run it against packages, you can pick these here:

https://teamcity.starcounter.org/viewLog.html?buildId=108028&buildTypeId=Bifrost_UmbrellaNovaBuildRelease

With that, we have test resources in place (I've merged them to `develop`, and then merged `develop` into `develop-inner-transact`) and test we have right now are green. At least something. 😄 What we need now, I guess, is better test coverage.

**Comment by per-samuelsson** [comment]
> It's not the same regarding synchronous exceptions `action` may throw: [#7 (comment)](https://github.com/Starcounter/Starcounter.Database.Extensions/pull/7#discussion_r397942130)

Switched to `Task.FromException` now. Good?

**Comment by bigwad** [comment]
>Switched to Task.FromException now. Good?

I wonder why to convert the methods to no-async if no-async code is more verbose.
What if Rollback also throws? The maintenance of such a method is an annoyance. I think that practical solution would be either (1) a top-level try/catch or (2) give up on asynchronous exceptions and let the methods throw synchronously or (3) put back async/await.

**Comment by per-samuelsson** [comment]
> I think that practical solution would be either (1) a top-level try/catch or (2) give up on asynchronous exceptions and let the methods throw synchronously or (3) put back async/await.

I vote (3). In 0810a3861665341401c8b76fbf8e109418274756, I actually made them `async` in the first place, to address this. I'm not sure why it was discarded later.

**Comment by per-samuelsson** [comment]
> > I think that practical solution would be either (1) a top-level try/catch or (2) give up on asynchronous exceptions and let the methods throw synchronously or (3) put back async/await.
> 
> I vote (3). In [0810a38](https://github.com/Starcounter/Starcounter.Database.Extensions/commit/0810a3861665341401c8b76fbf8e109418274756), I actually made them `async` in the first place, to address this. I'm not sure why it was discarded later.

What about this latest submission @bigwad?

**Comment by bigwad** [comment]
>What about this latest submission @bigwad?

It basically rollbacks the changes made in 39b7521aa272a0407ad74b7ba77667d5e33351de. I guess it's fine, but two non-async `TransactAsync` overloads are still there.

**Comment by per-samuelsson** [comment]
> > What about this latest submission @bigwad?
> 
> It basically rollbacks the changes made in [39b7521](https://github.com/Starcounter/Starcounter.Database.Extensions/commit/39b7521aa272a0407ad74b7ba77667d5e33351de). 

Well, those changes reverted those we did in https://github.com/Starcounter/Starcounter.Database.Extensions/commit/0810a3861665341401c8b76fbf8e109418274756, and I still didn't get why.

> I guess it's fine, but two non-async `TransactAsync` overloads are still there.

We didn't address those in https://github.com/Starcounter/Starcounter.Database.Extensions/commit/0810a3861665341401c8b76fbf8e109418274756, and as I understood it based on discussion around [here](https://github.com/Starcounter/Starcounter.Database.Extensions/pull/7#discussion_r397908722), it wasn't of any real importance. But I can try making them async in the same manner, sure.

**Comment by per-samuelsson** [comment]
> But I can try making them async in the same manner, sure.

And done. Hope it make sense.

File: src/Starcounter.Database.Extensions/NestedTransactor.cs [hunk 1]
```diff
@@ -114,47 +114,52 @@ public override async Task<T> TransactAsync<T>(Func<IDatabaseContext, T> functio
...
+            else
             {
-                Rollback(context);
-                return Task.FromException(e);
+                var context = new NestedTransactionContext(nested);
+
+                try
+                {
+                    await function(context);
+                    return;
```
Comment (bigwad) [inline src/Starcounter.Database.Extensions/NestedTransactor.cs]: not needed `return`

Comment (per-samuelsson) [reply src/Starcounter.Database.Extensions/NestedTransactor.cs]: Ha, that one I just dropped locally _before_ you mentioned it. :)

File: samples/NestedTransactions/Program.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                var jane = db.Insert<Person>();
+                jane.Name = "Jane Doe";
+                
+                transactor.Transact(db =>
+                {
+                    // Without NestedTransactor, this would not be allowed.
+                    // Starcounter natively does not support nested transactions and
+                    // will raise an exception for any Transact done as part of an
+                    // already running transaction.
```
Comment (bigwad) [inline samples/NestedTransactions/Program.cs]: Obsoleted due to recent changes

**Comment by per-samuelsson** [comment]
Merged as part of https://github.com/Starcounter/Starcounter.Database.Extensions/pull/7

**Pull request merged**